include/alpaka/queue/IGenericQueue.hpp

```diff
@@ -0,0 +1,57 @@
+/* Copyright 2019 Axel Huebl, Benjamin Worpitz, Matthias Werner
+ *
+ * This file is part of Alpaka.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#pragma once
+
+#include <alpaka/core/BoostPredef.hpp>
+
+// #include <alpaka/event/EventGeneric.hpp>
+// #include <alpaka/dev/DevCpu.hpp>
+
+namespace alpaka
+{
+    namespace event
+    {
+        template<typename TDev>
+        class EventGeneric;
+    }
+}
+
+namespace alpaka
+{
+    namespace queue
+    {
+#if BOOST_COMP_CLANG
+    // avoid diagnostic warning: "has no out-of-line virtual method definitions; its vtable will be emitted in every translation unit [-Werror,-Wweak-vtables]"
+    // https://stackoverflow.com/a/29288300
+    #pragma clang diagnostic push
+    #pragma clang diagnostic ignored "-Wweak-vtables"
+#endif
+
+        //#############################################################################
+        //! The CPU queue interface
+        template<
+            typename TDev>
```

Comment:
Could we make `IGenericQueue` depend on `TEvent` instead of `TDev`? The queue does not need to know its device at all but only the Event. `event::EventGeneric<TDev>` can be provided at the place the template is instantiated.

Reply:
I did it this way to ensure the template argument will always be derived from EventGeneric and exclude arbitrary types. If we do not want to be this strict, we could use `template<typename TEvent>` here. What shall it be?

Reply:
I do not see a reason to be so strict. Maybe there will be another queue that uses a totally different event.

Reply:
But IGenericThreadsQueue is a polymorphic interface for QueueThreadsGeneric* which uses EventGeneric<TDev>.

Allowing IGenericThreadsQueue to be used with other queues than instances of QueueThreadsGeneric* would extent the original scope of ICpuQueue. If we extent the scope this far, IGenericThreadsQueue could also be used for something like the Cuda queue, and thus would be restricted to living on the host anymore.

Reply:
At least this is what I understood the intent behind ICpuQueue to be: a polymorphic interface for queues using EventCpu.

Reply:
I am also ok with this. My point was that there might be another type of event that can be used with the `IGenericThreadsQueue` that is still compatible with the current CPU queue and device. but is not necessarily a specialization of `EventGeneric`.

Reply:
There can be no event which is not a specialization of `EventGenericThreads`, that can also be used with `QueueGenericThreats*` and by extension `QueueCpu*`, because the queue explictly uses `EventGenericThreads<TDev>` and also overrides `enqueue()` and `wait()` using this argument type. So if `IGenericThreadsQueue` were not restricted one of the following would need to change:
* formally allow `IGenericThreadsQueue` to be a base for different queue types
* add another template parameter `TEvent` to `QueueGeneric`, one could still enforce that `TEvent` and `TDev` are connected by saying
  ```c++
  template < typename TDev, template<typename TTDev> TEvent = EventGenericThreads >
  class QueueGenericThreats...
  ```
At the moment I would prefer to leave it strict. There is the option of relaxing it should the need arise, as `IGenericThreadsQueue` is an implementation detail from the perspective of user code.